iroha_config/src/lib.rs

```diff
@@ -0,0 +1,131 @@
+    #[derive(Debug)]
+    pub enum Error {
+        /// Got unknown field
+        UnknownField(Vec<String>),
+        /// Failed to deserialize or serialize field
+        FieldError(FieldError),
+    }
+
+    impl fmt::Display for Error {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            match self {
+                Self::UnknownField(field) => {
+                    let field = field
+                        .iter()
+                        .skip(1)
+                        .fold(field[0].clone(), |prev, suc| prev + "." + suc);
+                    write!(f, "Failed to deserialize: Unknown field {}", field)
+                }
+                Self::FieldError(_) => write!(f, "Failed to deserialize"),
+            }
+        }
+    }
+
+    impl StdError for Error {
+        fn source(&self) -> Option<&(dyn StdError + 'static)> {
+            if let Error::FieldError(ref field) = self {
+                Some(field)
+            } else {
+                None
+            }
+        }
+    }
+
+    impl Error {
+        /// Constructs field error
+        pub const fn field_error(field: &'static str, error: serde_json::Error) -> Self {
+            Self::FieldError(FieldError { field, error })
+        }
+    }
+}
+
+/// Trait for dynamic configuration via maintanence endpoint for rust structures
+pub trait Configurable: Serialize + DeserializeOwned {
+    /// Error type returned by methods of trait
+    type Error;
+
+    /// Sets field of structure with json-value
+    /// # Errors
+    /// Fails if field was unknown or if fails Deserialize value
+    fn set(&mut self, field: &str, value: impl Into<Value>) -> Result<(), Self::Error> {
+        self.set_recursive([field], value)
+    }
+
+    /// Sets inner field of arbitrary inner depth with json-value
+    /// # Errors
+    /// Fails if field was unknown or if fails Deserialize value
+    fn set_recursive<'a>(
+        &mut self,
+        inner_field: impl AsRef<[&'a str]>,
+        value: impl Into<Value>,
+    ) -> Result<(), Self::Error>;
+
+    /// Gets field of structure and returns as json-value
+    /// # Errors
+    /// Fails if field was unknown
+    fn get(&self, field: &str) -> Result<Value, Self::Error> {
+        self.get_recursive([field])
+    }
+
+    /// Gets inner field of arbitrary inner depth and returns as json-value
+    /// # Errors
+    /// Fails if field was unknown
+    fn get_recursive<'a>(&self, inner_field: impl AsRef<[&'a str]>) -> Result<Value, Self::Error>;
+
+    /// Fails if fails to deserialize from environment
+    /// # Errors
+    /// Fails if fails to deserialize from environment
+    fn load_environment(&mut self) -> Result<(), Self::Error>;
+
+    /// Gets docs of inner field of arbitrary depth
+    /// # Errors
+    /// Fails if field was unknown
+    fn get_doc_recursive<'a>(
+        &self,
+        inner_field: impl AsRef<[&'a str]>,
+    ) -> Result<Option<&'static str>, Self::Error>;
```

Comment:
Can you explain please that is a benefits of returning of `Result<Option<T>, Err>` instead of just `Result<T, Err>`

Reply:
Because if there are no doc strings for field it is not actually an error. This is a valid state and we need to tell from our endpoint that there are no documentation here.

There could be other design there some of error variant should tell that there are no docs for field, but then we will obligate user to have such variant (as i said that doesn't seem to be an actual error) and signature then will show that for any valid field we will have docs.